src/node_crypto.cc

```diff
@@ -3190,16 +3201,102 @@ void CipherBase::InitIv(const FunctionCallbackInfo<Value>& args) {
     iv_buf = Buffer::Data(args[2]);
     iv_len = Buffer::Length(args[2]);
   }
-  cipher->InitIv(*cipher_type, key_buf, key_len, iv_buf, iv_len);
+  CHECK(args[3]->IsInt32());
+  // Don't assign to cipher->auth_tag_len_ directly; the value might not
+  // represent a valid length at this point.
+  int auth_tag_len = args[3]->Int32Value();
+
+  cipher->InitIv(*cipher_type, key_buf, key_len, iv_buf, iv_len, auth_tag_len);
+}
+
+
+bool CipherBase::InitAuthenticated(const char *cipher_type, int iv_len,
+                                   int auth_tag_len) {
+  CHECK(IsAuthenticatedMode());
+
+  // TODO(tniessen) Use EVP_CTRL_AEAD_SET_IVLEN when migrating to OpenSSL 1.1.0
+  static_assert(EVP_CTRL_CCM_SET_IVLEN == EVP_CTRL_GCM_SET_IVLEN,
+                "OpenSSL constants differ between GCM and CCM");
+  if (!EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_GCM_SET_IVLEN, iv_len, nullptr)) {
+    env()->ThrowError("Invalid IV length");
+    return false;
+  }
+
+  if (EVP_CIPHER_CTX_mode(ctx_) == EVP_CIPH_CCM_MODE) {
+    if (auth_tag_len < 0) {
+      char msg[128];
+      snprintf(msg, sizeof(msg), "authTagLength required for %s", cipher_type);
+      env()->ThrowError(msg);
+      return false;
+    }
+
+#ifdef NODE_FIPS_MODE
+    // TODO(tniessen) Support CCM decryption in FIPS mode
+    if (kind_ == kDecipher && FIPS_mode()) {
+      env()->ThrowError("CCM decryption not supported in FIPS mode");
+      return false;
+    }
+#endif
+
+    if (!EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_CCM_SET_TAG, auth_tag_len,
+        nullptr)) {
+      env()->ThrowError("Invalid authentication tag length");
+      return false;
+    }
+
+    // When decrypting in CCM mode, this field will be set in setAuthTag().
+    if (kind_ == kCipher)
+      auth_tag_len_ = auth_tag_len;
+
+    // The message length is restricted to 2 ^ (15 - iv_len) - 1 bytes.
```

Comment:
Is there an `8 *` multiplier missing here in the exponent?

Reply:
Oh yeah sure!